sync-core/src/main/java/com/cloudant/http/CookieFilter.java

```diff
@@ -0,0 +1,147 @@
+    final String sessionRequestBody;
+    private String cookie = null;
+    private boolean shouldAttemptCookieRequest = true;
+
+    /**
+     * Constructs a cookie filter.
+     * @param username The username to use when getting the cookie
+     * @param password The password to use when getting the cookie
+     */
+    public CookieFilter(String username, String password){
+        Map<String,String> sessionRequestMap = new HashMap<String, String>();
+        sessionRequestMap.put("name", username);
+        sessionRequestMap.put("password", password);
+        JSONHelper helper = new JSONHelper();
+        sessionRequestBody = helper.toJson(sessionRequestMap);
+    }
+
+
+    @Override
+    public HttpConnectionFilterContext filterRequest(HttpConnectionFilterContext context) {
+
+        HttpURLConnection connection = context.connection.getConnection();
+
+        if(shouldAttemptCookieRequest) {
+            if (cookie == null) {
+                cookie = getCookie(connection.getURL());
+            }
+            connection.setRequestProperty("Cookie", cookie);
+        }
+
+        return context;
+
+    }
+
+    @Override
+    public HttpConnectionFilterContext filterResponse(HttpConnectionFilterContext context) {
+        HttpURLConnection connection = context.connection.getConnection();
+        if (context.connection.getResponseCode() == 401) {
+            //we need to get a new cookie
+            cookie = getCookie(connection.getURL());
+            //don't resend request, failed to get cookie
+            if(cookie != null) {
+                context.replayRequest = true;
+
+                context = new HttpConnectionFilterContext(context);
+            } else {
+                context.replayRequest = false;
+            }
+        }
+        return context;
+
+    }
+
+     private String getCookie(URL url){
+
+        try {
+            URL sessionURL = new URL(String.format("%s://%s:%d/_session",
+                    url.getProtocol(),
+                    url.getHost(),
+                    url.getPort()));
+
+            HttpConnection conn = Http.POST(sessionURL, "application/json");
+            conn.setRequestBody(sessionRequestBody);
+            HttpURLConnection connection = conn.execute().getConnection();
+            String cookieHeader = connection.getHeaderField("Set-Cookie");
+            int responseCode = connection.getResponseCode();
+
+
+            if(responseCode == 401){
+                shouldAttemptCookieRequest  = false;
+                logger.severe("Credentials are incorrect, cookie authentication will not be attempted again");
+            } else if (responseCode / 100 == 5){
+                logger.log(Level.SEVERE,
+                        "Failed to get cookie from server, response code %s, cookie auth",
+                        responseCode);
+            } else if(responseCode / 100 == 2) {
+                return cookieHeader.substring(0, cookieHeader.indexOf(";"));
```

Comment:
Such as? Any proxying the user is doing should be transparent, other parts of the code base would  likely break if it wasn't.